src/sage/modules/free_module.py

```diff
@@ -6685,8 +6721,17 @@ def __init__(self, ambient, basis, check=True,
                 raise TypeError("each element of basis must be in "
                                 "the ambient vector space")
 
-        if echelonize and not already_echelonized:
-            basis = self._echelonized_basis(ambient, basis)
+        basis = basis_seq(V, basis)
+
+        # normalize the parameters. We have (basis, echelonize, echelonized_basis, already_echelonized)
+        if already_echelonized:
+            assert echelonized_basis is None or echelonized_basis == basis, "inconsistent parameters provided"
```

Comment:
```suggestion
            assert echelonized_basis is None, "inconsistent parameters provided"
```
I don't understand the point of the second test. The basis given could be the echelon basis. I don't see anything inconsistent with doing that (yes, the user was being redundant, but I don't see any problem with having this situation).

Reply:
isn't that backward?

* with the current code, if the user being redundant, the program will ignore it.
* with your suggestion, if the user specify `basis=X, echelonized_basis=X`, it will error out.

Reply:
Sorry, I forgot that this is an `assert` and not checking validity of input. We shouldn't use `assert` for validating user input like this (this is a general policy that might not be explicitly written down anywhere).

That said, the other point remains valid, I don't think the basis doesn't need to be the echelonized basis; see my other comment. (I also don't see how my change here causes that errors out either.)

Reply:
I'd say the whole code feels brittle also, although I can't say I fully understand what it does (and throwing the whole thing away and reimplement it is likely to introduce hundreds of test failures), so for safety I try to make minimal changes to get the tests pass.

No, the point is the following. The `already_echelonized` parameter is meant to be used like this (looking at the current doctest):

```
            sage: W = FreeModule_submodule_with_basis_pid(V, [[1,2,0], [0,0,1]], already_echelonized=True)
```

in other words, if `already_echelonized=True` is passed, the user expects to pass the echelonized basis **in the `basis` argument**.

If the user passes different `basis` and `echelonized_basis`, and declare that both of them are echelonized, it must be an user mistake.

so, (yes, a lot of logical thinking needed here, indeed, but the API was like that):

* if the user call `ModuleClass(basis, already_echelonized=True)`, we expect `basis` to be already echelonized
* if the user call `ModuleClass(basis, echelonized_basis=basis, already_echelonized=True)`, should we error out or silently accept? Currently I silently accept but you suggest to error out.

Furthermore, on this comment

> It also does a subtle change since it sets the basis to the echelonized basis (so it is not a no-op). In principle, the user might want the module to hold onto this

That's required by the external API (again, looking at the current doctest):

```
            sage: W = FreeModule_submodule_with_basis_pid(V, [[1,2,3], [1,1,1]], echelonize=True)
            sage: W._FreeModule_submodule_with_basis_pid__echelonized_basis_matrix is not None
            True
            sage: W.echelonized_basis_matrix()
            [ 1  0 -1]
            [ 0  1  2]
```

when the user call `ModuleClass(basis, echelonize=True)`, **the user expect to use `echelonize(basis)` as the distinguished basis**. So we must echelonize it here.

Seriously, I don't see a value of having a category of objects with a **distinguished** basis, but that is what the API is at this point and I feel modifying it would require too much work.

(tangent: nor do I see the value of having a category of objects `Sets().Enumerated()` with a **distinguished** enumeration order, rather than just being countable (i.e. have cardinality at most $|ℕ|$), and it is completely arbitrary that currently ℤ is in Sets().Enumerated() but ℚ is not. Anyway this is not related to this pull request)

Reply:
I agree that this is clunky. I am not sure either what is the relevance of all this clunkiness and several provided parameters but I agree that if we feel the user is specifying too many parameters to the API then should be provided, the API should error out.

Reply:
And it is really mind boggling, trying to understand what goes on here but we can make a note (a `TODO` if you wish) to simplify the API if there are no usecases.